### PR TITLE
executor: open childExec during execution for UnionExec

### LINKED ISSUE #21441: Support lazy open union children
## Development Task

In #19827, we support concurrency control for  Union's children execution.

but Union executor still `Init` all children before execute `Next`

init children will cause sub-plan nodes's `Init` be called

combined with #14320, partition table will make the "buildCopTasks takes too much time" problem even worse.

so for a partition table with 128 partitions

```
select * from partition_table limit 1;
```
it will build all copTask in memory for each 128 partitions before send any request to tikv(and finally maybe only 1 request is useful)

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #21441 

re-revert https://github.com/pingcap/tidb/pull/21561

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
Before this commit, `Union` opens all the child executors during the `Open` phase.
This increases the overhead of `buildRequest` and the risk of OOM.

After this commit, the `Open` of childExec in `Union` is moved to the `Next` phase. The childExec be opened simultaneously is limited to `tidb_executor_concurrency` atmost.

How it Works:

### Related changes

N/A

### Check List 

Tests 

- Integration test
- Manual test
Suppose we have a table `t`, it contains 100 partitions, each partition contains 10,000,000 tuples.
The result of `select * from t;`
Before:
image
After:
The query works fine.

Side effects

- The speed of UnionExec may decrease if the child of Union is larger than `tidb_executor_concurrency`.

### Release note 

- Refine the execution mechanism `UnionExec` to reduce the overhead of `UnionExec`.

### CODE REVIEW

**Review by wshwsh12 — APPROVED** [review]

**Review by ichn-hu — COMMENTED** [review]
Rest LGTM

File: executor/executor.go [hunk 1]
```diff
@@ -1600,7 +1626,15 @@ func (e *UnionExec) Close() error {
 		for range e.childIDChan {
 		}
 	}
-	return e.baseExecutor.Close()
+	// We do not need to acquire the e.mu.Lock since all the resultPuller can be
+	// promised to exit when reaching here (e.childIDChan been closed).
+	var firstErr error
+	for i := 0; i <= e.mu.maxOpenedChildID; i++ {
+		if err := e.children[i].Close(); err != nil && firstErr == nil {
```
[the remark this answers is not in the document]
Comment (ichn-hu) [reply executor/executor.go:1633]: @XuHuaiyu PTAL

[the remark this answers is not in the document]
Comment (guo-shaoge) [reply executor/executor.go:1633]: Looks like `maxOpenedChildID` can guarantee all children is opend.

**Review by ichn-hu — APPROVED** [review]
LGTM

**Pull request merged**